Fix time ago when making a notification read/unread.

ralinc opened the pull request:
Marking a notification as unread resets the timeago stamp causing
the times to look wrong. It can be reproduced by marking an old
notification as unread. Using the update_colum instead of
update_attribute will not touch the updated_at attribute, and
thus will not affect the updated time ago in the view.

Fixes #6798.

SuperTux88 commented on app/models/notification.rb:
```diff
@@ -41,7 +41,11 @@ def email_the_user(target, actor)
 
   def set_read_state( read_state )
-    self.update_attributes( :unread => !read_state )
+    if self.new_record?
```
This method is only called on existing notifications, so this `if` shouldn't be needed.

ralinchimev replied:
The if-statement is needed because of some failing model unit tests. Or we have to change the tests somehow.

SuperTux88 replied:
We shouldn't change the code only to make tests running. Please change the tests that they use saved notifications, so the tests correspond to the real conditions.

ralinchimev replied:
Of course. I agree.

SuperTux88 commented on spec/controllers/notifications_controller_spec.rb:
```diff
@@ -31,6 +31,14 @@
       get :update, "id" => note.id, :set_unread => "true", :format => :json
     end
 
+    it 'marks a notification as unread without timestamping' do
```
You should use `TimeCop` here to change the time. Because this spec runs in the same second as the Notification was created, so the time would be the same anyway.

Also this spec is red in travis, because after save it to the database some milliseconds were truncated. So you need to compare only date, hours, minutes and seconds (or only the one you changed with `TimeCop`).

ralinchimev replied:
Done. Thanks for the feedback.

diaspora-code-review commented on app/models/notification.rb:
```diff
@@ -41,7 +41,11 @@ def email_the_user(target, actor)
   def set_read_state( read_state )
-    self.update_attributes( :unread => !read_state )
+    if self.new_record?
+      self.update_attribute(:unread, !read_state)
```
Redundant `self` detected.

diaspora-code-review commented on app/models/notification.rb:
```diff
@@ -41,7 +41,11 @@ def email_the_user(target, actor)
+    if self.new_record?
+      self.update_attribute(:unread, !read_state)
+    else
+      self.update_column(:unread, !read_state)
```
Redundant `self` detected.

diaspora-code-review commented on spec/controllers/notifications_controller_spec.rb:
```diff
@@ -31,6 +31,19 @@
       get :update, "id" => note.id, :set_unread => "true", :format => :json
     end
 
+    it 'marks a notification as unread without timestamping' do
```
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

ralinc committed:
Do not change notification timestamp when making it read or unread.

Marking a notification as unread resets the timeago stamp causing
the times to look wrong. It can be reproduced by marking an old
notification as unread. Using the update_column instead of
update_attribute will not touch the updated_at attribute, and
thus will not affect the updated time ago in the view.

Fixes #6798.

denschub commented:
Merged as 1773e3e35d3cf706a06088d4c36155a5ff4ab916, thank you for your contribution!

Flaburgan commented:
Thank you!
